### PR TITLE
ICSAAS-308/312/314-update search component with bestbets autosuggest and autocomplete

### PR DESCRIPTION
**Pull request by igorsimovski**

### CODE REVIEW

File: api/src/main/java/com/valtech/aem/saas/api/fulltextsearch/FulltextSearchService.java [hunk 1]
```diff
@@ -1,19 +1,29 @@
 package com.valtech.aem.saas.api.fulltextsearch;
 
+/**
+ *
```
Comment (nhirrle) [inline api/src/main/java/com/valtech/aem/saas/api/fulltextsearch/FulltextSearchService.java:4]: - [ ] please delete empty comment

File: core/src/main/java/com/valtech/aem/saas/core/autocomplete/AutocompleteJsonResponse.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  @PostConstruct
+  private void init() {
+    if (!response.isCommitted()) {
+      Optional.ofNullable(request.adaptTo(RequestWrapper.class))
+          .ifPresent(requestWrapper ->
+              requestWrapper.getParameter(SearchResultsImpl.SEARCH_TERM).ifPresent(text -> {
+                SearchConfiguration searchConfiguration = request.getResource().adaptTo(ConfigurationBuilder.class)
+                    .as(SearchConfiguration.class);
+                if (StringUtils.isNotBlank(searchConfiguration.index())) {
+                  TypeaheadPayload payload = DefaultTypeaheadPayload.builder()
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/autocomplete/AutocompleteJsonResponse.java:59]: we will need to adopt here to the search module as we will want to apply same pre-filters as in the search component defined

File: core/src/main/java/com/valtech/aem/saas/core/common/response/JsonResponseFlusher.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import java.nio.charset.StandardCharsets;
+import java.util.function.Consumer;
+import javax.servlet.http.HttpServletResponse;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.sling.api.SlingHttpServletResponse;
+import org.apache.sling.servlets.post.JSONResponse;
+
+@Slf4j
+public class JsonResponseFlusher {
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/common/response/JsonResponseFlusher.java:14]: - [ ] for all classes and public methods always add javadoc

File: core/src/main/java/com/valtech/aem/saas/core/common/saas/SaasIndexValidator.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package com.valtech.aem.saas.core.common.saas;
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+import org.apache.commons.lang3.StringUtils;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class SaasIndexValidator {
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/common/saas/SaasIndexValidator.java:8]: ideally we can get rid of this, see comment in AutocompleteJsonResponse

File: core/src/main/java/com/valtech/aem/saas/core/fulltextsearch/SearchImpl.java [hunk 5]
```diff
@@ -46,15 +57,15 @@
 public class SearchImpl implements Search {
 
   public static final String RESOURCE_TYPE = "saas-aem-module/components/search";
-  public static final String NN_SEARCH_RESULTS_TABS_CONTAINER = "searchresults-tabs";
+  public static final String NN_SEARCH_TABS_CONTAINER = "search-tabs";
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/fulltextsearch/SearchImpl.java:60]: - [ ] NN = Node Name? if so, please rename to NODE_NAME_SEARCH_TABS_CONTAINER

File: core/src/main/java/com/valtech/aem/saas/core/fulltextsearch/SearchImpl.java [hunk 6]
```diff
@@ -133,4 +142,51 @@ public String getLoadMoreButtonText() {
...
+        .map(Arrays::stream)
+        .orElse(Stream.empty())
+        .map(searchFilterConfiguration -> new FilterImpl(searchFilterConfiguration.name(),
+            searchFilterConfiguration.value()))
+        .collect(Collectors.toList());
+  }
+
+  private String getSearchTabUrl(@NonNull Resource searchTab) {
+    try {
+      return new URIBuilder(String.format("%s.%s.%s", searchTab.getPath(), ExporterConstants.SLING_MODEL_SELECTOR,
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/fulltextsearch/SearchImpl.java:185]: this doesn't work with sling mappings, so might only work on author. 
we need to clarify how we want to do resource mapping

File: core/src/main/java/com/valtech/aem/saas/core/fulltextsearch/SearchResultsImpl.java [hunk 7]
```diff
@@ -62,85 +76,149 @@
...
 
-  private int resolveResultsPerPage(RequestWrapper requestWrapper) {
+  private int getStartPage(RequestWrapper requestWrapper) {
+    return requestWrapper.getParameter(QUERY_PARAM_START)
+        .map(start -> new StringToInteger(start).asInt())
+        .map(OptionalInt::getAsInt)
+        .orElse(DEFAULT_START_PAGE);
+  }
+
+  private void initSearch(String searchTerm, RequestWrapper requestWrapper) {
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/fulltextsearch/SearchResultsImpl.java:158]: that method is pretty big still. can  you introduce few wrapper functions, e.g. for initializing fulltextSearchGetRequestPayload  ?

File: core/src/main/java/com/valtech/aem/saas/core/http/response/SuggestionDataExtractionStrategy.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * A strategy for extracting highlighting data.
+ */
+@Slf4j
+public final class SuggestionDataExtractionStrategy implements SearchResponseDataExtractionStrategy<Suggestion> {
+
+  public static final String PN_SPELLCHECK = "spellcheck";
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/http/response/SuggestionDataExtractionStrategy.java:16]: - [ ] what does PN stand for? Property Name? can you replace without using abbrevations?

File: core/src/main/java/com/valtech/aem/saas/core/typeahead/DefaultTypeaheadService.java [hunk 9]
```diff
@@ -88,17 +86,17 @@ private void activate(Configuration configuration) {
   public @interface Configuration {
 
     String DEFAULT_API_ACTION = "/typeahead";
-    String DEFAULT_API_VERSION_PATH = "/api/v3";
+    String DEFAULT_API_VERSION_PATH = "/api/v3"; //NOSONAR
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/typeahead/DefaultTypeaheadService.java:89]: just a small one_ Please add a space after comments

File: ui.apps/src/main/content/jcr_root/apps/saas-aem-module/shared/filters.xml [hunk 10]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<jcr:root
+    xmlns:sling="http://sling.apache.org/jcr/sling/1.0"
+    xmlns:jcr="http://www.jcp.org/jcr/1.0"
+    xmlns:nt="http://www.jcp.org/jcr/nt/1.0"
+    jcr:primaryType="nt:unstructured"
+    fieldLabel="Filters"
+    fieldDescription="Only simple filter entries can be configured."
```
Comment (nhirrle) [inline ui.apps/src/main/content/jcr_root/apps/saas-aem-module/shared/filters.xml:8]: please replace description with "Filter entries with the "AND" relation can be configured"

File: core/src/main/java/com/valtech/aem/saas/core/fulltextsearch/DefaultFulltextSearchService.java [hunk 11]
```diff
@@ -50,71 +57,85 @@ public int getRowsMaxLimit() {
...
+      }
+      FulltextSearchResults.FulltextSearchResultsBuilder fulltextSearchResultsBuilder =
+          FulltextSearchResults.builder()
+              .totalResultsFound(responseBody.get().getNumFound())
+              .currentResultPage(responseBody.get().getStart())
+              .results(results.collect(Collectors.toList()));
+      if (enableAutoSuggest) {
+        log.debug("Auto suggest is enabled.");
+        searchResponse.get(new SuggestionDataExtractionStrategy()).flatMap(suggestion -> LoggedOptional.of(suggestion,
+                logger -> logger.warn("No suggestion has been found in search response")))
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/fulltextsearch/DefaultFulltextSearchService.java:101]: not sure if we should log this on warn level. it is very likely that we don't get a suggestion as users might input any value. can you change to debug level instead?

File: core/src/main/java/com/valtech/aem/saas/core/fulltextsearch/SearchImpl.java (hunk 6 above)
Comment (nhirrle) [reply core/src/main/java/com/valtech/aem/saas/core/fulltextsearch/SearchImpl.java]: - [ ] lets tackle this today

File: api/src/main/java/com/valtech/aem/saas/api/caconfig/SearchConfiguration.java [hunk 12]
```diff
@@ -7,16 +7,30 @@
...
-  @Property(label = "Search Templates", description = "List of custom query templates' names, for specialized/different field boosting strategies. (Optional)")
+  @Property(label = "Search Templates",
+      description = "List of custom query templates' names, for specialized/different field boosting strategies. (Optional)")
   String[] templates() default {};
 
+  @Property(label = "Enable Best Bets",
+      description = "Flag that enables displaying best bets on the top of the search results.")
+  boolean enableBestBets() default DEFAULT_ENABLE_BEST_BETS;
+
+  @Property(label = "Enable Best Bets",
```
[the remark this answers is not in the document]
Comment (igorsimovski) [reply api/src/main/java/com/valtech/aem/saas/api/caconfig/SearchConfiguration.java]: fixed, thx :)

File: core/src/main/java/com/valtech/aem/saas/core/indexing/PathTransformerPipeline.java [hunk 13]
```diff
@@ -1,11 +1,12 @@
...
+import org.apache.sling.api.SlingHttpServletRequest;
 
 /**
  * Represents a service that applies one or more externalization rules to a path input in a pipeline manner.
  */
-public interface PathExternalizerPipeline {
+public interface PathTransformerPipeline {
 
   /**
    * Gets list of paths as product od the applied externalization rules.
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/indexing/PathTransformerPipeline.java:12]: typo

File: core/src/main/java/com/valtech/aem/saas/core/indexing/PathTransformerPipeline.java [hunk 14]
```diff
@@ -14,4 +15,6 @@
    * @return list of externalized paths.
    */
   List<String> getExternalizedPaths(String path);
+
+  String getMappedPath(SlingHttpServletRequest request, String path);
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/indexing/PathTransformerPipeline.java:19]: - [ ] javadoc please

File: core/src/main/java/com/valtech/aem/saas/core/indexing/PathTransformerPipelineService.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicyOption;
+
+@Slf4j
+@Component(name = "Search as a Service - Path Transformer Pipeline",
+    service = PathTransformerPipeline.class)
+public class PathTransformerPipelineService implements PathTransformerPipeline {
+
+  @Reference(cardinality = ReferenceCardinality.MULTIPLE, policyOption = ReferencePolicyOption.GREEDY)
+  private List<PathTransformer> pathTransformers = new CopyOnWriteArrayList<>();
```
Comment (nhirrle) [inline core/src/main/java/com/valtech/aem/saas/core/indexing/PathTransformerPipelineService.java:20]: do we have a default pathTransformer? we need to have 1 which uses DayCQExternalizer

**Pull request merged**